File: backend/src/main/java/com/skewwhiffy/auraltester/notation/model/note/NoteLength.java
package com.skewwhiffy.auraltester.notation.model.note;

import com.skewwhiffy.auraltester.fraction.Fraction;

public record NoteLength(Fraction length) {
    public NoteLength(int length) {
        this(new Fraction(length, 1));
    }
    /*
  val dotted
    get() = NoteLength((Fraction(3, 2) * length))
     */

    public String getAbc() {
        return length.getTopHeavyString();
    }

    public static NoteLength getBreve() {
        return new NoteLength(2);
    }

    public static NoteLength getSemibreve() {
        return new NoteLength(1);
    }

    public static NoteLength getMinim() {
        return new NoteLength(new Fraction(1, 2));
    }

    public static NoteLength getCrotchet() {
        return new NoteLength(new Fraction(1, 4));
    }

    public static NoteLength getQuaver() {
        return new NoteLength(new Fraction(1, 8));
    }

    public static NoteLength getSemiquaver() {
        return new NoteLength(new Fraction(1, 16));
    }

    public static NoteLength getDemisemiquaver() {
        return new NoteLength(new Fraction(1, 32));
    }

    public static NoteLength getHemidemisemiquaver() {
        return new NoteLength(new Fraction(1, 64));
    }

}

File: backend/src/main/kotlin/com/skewwhiffy/auraltester/notation/model/note/NoteLength.kt
package com.skewwhiffy.auraltester.notation.model.note

import com.skewwhiffy.auraltester.fraction.Fraction

data class NoteLength(val length: Fraction) {
    constructor(length: Int): this(Fraction(length))

    companion object {
        val breve: NoteLength
            get() = NoteLength(2)

        val semibreve: NoteLength
            get() = NoteLength(1)

        val minim: NoteLength
            get() = NoteLength(Fraction(1, 2))

        val crotchet: NoteLength
            get() = NoteLength(Fraction(1, 4))

        val quaver: NoteLength
            get() = NoteLength(Fraction(1, 8))

        val semiquaver: NoteLength
            get() = NoteLength(Fraction(1, 16))

        val demisemiquaver: NoteLength
            get() = NoteLength(Fraction(1, 32))

        val hemidemisemiquaver: NoteLength
            get() = NoteLength(Fraction(1, 64))

    }

    val dotted by lazy { NoteLength(Fraction(3, 2) * length) }

    val abc: String
        get() = length.topHeavyString
}
